### PR TITLE
IN PROGRESS: React-query implemented + axios 

### PR DESCRIPTION
**Pull request by min-in-zon**
#39

### CODE REVIEW

File: src/utilities/axios/axios.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import axios from 'axios';
+
+const instance = axios.create({
+  baseURL: import.meta.env.VITE_API_BASE_URL,
+  timeout: 1000,
```
Comment (adadychanka) [inline src/utilities/axios/axios.js:5]: No big deal, but I would increase it to 2-3 sec

File: src/services/PopularCitiesServise.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  const params = {
+    page,
+    pageSize,
+    sortOrder,
+    onlyWithGuides,
+  };
+
+  try {
+    const response = await instance.get(endpoint, { params });
+    if (response.data && response.data.data.items) {
```
Comment (adadychanka) [inline src/services/PopularCitiesServise.js:24]: Duplicated checks might be replaced with options chaining operator

```suggestion
    if (response?.data?.data?.items) {
```

File: src/services/PopularCitiesServise.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import instance from '../utilities/axios/axios';
+
+const DEFAULT_PAGE = 1;
+const DEFAULT_PAGE_SIZE = 10;
+const DEFAULT_SORT_ORDER = 'POPULARITY_DESC';
+const DEFAULT_ENDPOINT = '/cities';
+
+async function fetchPopularCities({
+  endpoint = DEFAULT_ENDPOINT,
```
Comment (adadychanka) [inline src/services/PopularCitiesServise.js:9]: hmmm, I don't think you need `endpoint` passed as a parameter here

File: src/services/PopularCitiesServise.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    sortOrder,
+    onlyWithGuides,
+  };
+
+  try {
+    const response = await instance.get(endpoint, { params });
+    if (response.data && response.data.data.items) {
+      return response.data.data.items;
+    } else {
+      throw new Error('No data found');
```
Comment (adadychanka) [inline src/services/PopularCitiesServise.js:27]: Be more specific, error message is too general

```suggestion
      throw new Error('Unable to retrieve popular cities');
```

File: src/main.jsx [hunk 5]
```diff
@@ -1,13 +1,23 @@
...
+} from '@tanstack/react-query';
+import { ReactQueryDevtools } from '@tanstack/react-query-devtools';
 import { BrowserRouter } from 'react-router-dom';
 import App from './App.jsx';
 import './index.scss';
 
+const queryClient = new QueryClient();
+
 ReactDOM.createRoot(document.getElementById('root')).render(
-  <React.StrictMode>
```
Comment (adadychanka) [inline src/main.jsx:8]: Why did you remove `StrictMode` ?

File: src/main.jsx [hunk 6]
```diff
@@ -1,13 +1,23 @@
 import React from 'react';
 import ReactDOM from 'react-dom/client';
+import {
+  useQuery,
+  useMutation,
+  useQueryClient,
+  QueryClient,
+  QueryClientProvider,
+} from '@tanstack/react-query';
+import { ReactQueryDevtools } from '@tanstack/react-query-devtools';
```
Comment (adadychanka) [inline src/main.jsx:10]: You imported dev tools but they haven't been used

https://tanstack.com/query/v4/docs/framework/react/devtools

File: src/main.jsx [hunk 7]
```diff
@@ -1,13 +1,23 @@
 import React from 'react';
 import ReactDOM from 'react-dom/client';
+import {
+  useQuery,
```
Comment (adadychanka) [inline src/main.jsx:4]: Removed unused imports
image

File: src/hooks/usePopularCities.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { useQuery } from '@tanstack/react-query';
+import fetchPopularCities from '../services/PopularCitiesServise';
+import transformCityItem from '../data/transformCityItem';
+
+const popularCitiesQueryKey = 'popularCities';
```
Comment (adadychanka) [inline src/hooks/usePopularCities.js:5]: let's follow constant naming convention we have chosen

```suggestion
const POPULAR_CITIES_QUERY_KEY = 'popularCities';
```

File: src/hooks/usePopularCities.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+const usePopularCities = () => {
+  const { data, error, isLoading } = useQuery({
+    queryKey: [popularCitiesQueryKey],
+    queryFn: fetchPopularCities,
+  });
+
+  const transformedData = data?.map(transformCityItem);
```
Comment (adadychanka) [inline src/hooks/usePopularCities.js:13]: `useMemo` is useful here to avoid array recreating during state changes hidden in `useQuery`

Comment (adadychanka) [reply src/hooks/usePopularCities.js:13]: + I usually add default value to avoid unexpected error outside the hook

```suggestion
  const transformedData = data?.map(transformCityItem) ?? [];
```

File: src/data/transformCityItem.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import Stub from '../assets/img/cities/Stub.png';
+
+const transformCityItem = city => {
+  return {
+    id: city.id,
+    name: city.name,
+    country: city.country.name,
```
Comment (adadychanka) [inline src/data/transformCityItem.js:7]: I would put optional chaining here in case BE is accidentally missing county for the city

```suggestion
    country: city.country?.name,
```

File: src/components/common/CitiesList/CitiesListSkeleton.jsx [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+            <div className="ml-1.5 mt-2.5 h-6 w-32 rounded bg-gray-300"></div>
+            <div className="ml-1.5 mt-2.5 h-4 w-24 rounded bg-gray-300"></div>
+          </div>
+        </li>
+      ))}
+    </ul>
+  );
+};
+
+export default CitiesListSkeleton;
```
Comment (adadychanka) [inline src/components/common/CitiesList/CitiesListSkeleton.jsx:19]: `memo`

File: src/components/common/CitiesList/CitiesList.jsx [hunk 12]
```diff
@@ -1,21 +1,34 @@
...
+
+const CitiesList = ({ limit, increaseLimit, showLoadMoreButton }) => {
+  const { data: cities, error, isLoading } = usePopularCities();
 
-const CitiesList = ({ cities, limit, increaseLimit, showLoadMoreButton }) => {
   const defaultLimit = 10;
   const citiesLimit = limit || defaultLimit;
-  const displayedCities = cities.slice(0, citiesLimit);
-  const allCitiesDisplayed = citiesLimit >= cities.length;
+  const displayedCities = cities?.slice(0, citiesLimit) || [];
```
Comment (adadychanka) [inline src/components/common/CitiesList/CitiesList.jsx:12]: having default value of data inside the hook will help to avoid all these fallbacks

Comment (adadychanka) [reply src/components/common/CitiesList/CitiesList.jsx:12]: and let's wrap `displayedCities` into `useMemo` to avoid getting a new array instance

File: src/main.jsx (hunk 5 above)
Comment (min-in-zon) [reply src/main.jsx:8]: In the video tutorial on React Query, some guy deleted it to remove repeated renderings. As far as I understand, it was not worth deleting strict mode..

File: src/services/PopularCitiesServise.js (hunk 4 above)
Comment (min-in-zon) [reply src/services/PopularCitiesServise.js:27]: When I caught an error, in order to understand at what stage of the request I was unable to receive data, I did console.log at each stage and wrote down all the errors briefly. You're right, I need to rename and add a couple more.

File: src/services/PopularCitiesServise.js (hunk 3 above)
Comment (min-in-zon) [reply src/services/PopularCitiesServise.js:9]: Who am I to argue

File: src/services/PopularCitiesServise.js (hunk 2 above)
Comment (min-in-zon) [reply src/services/PopularCitiesServise.js:24]: NEXT LVL 🔥🔥🔥

File: src/utilities/axios/axios.js (hunk 1 above)
Comment (min-in-zon) [reply src/utilities/axios/axios.js:5]: My wife tells me the same thing every night..

File: src/hooks/usePopularCities.js (hunk 8 above)
Comment (min-in-zon) [reply src/hooks/usePopularCities.js:5]: Do we use this naming only at the top level? Can we use a camel case in function blocks and etc?

File: src/components/common/CitiesList/CitiesListSkeleton.jsx (hunk 11 above)
Comment (min-in-zon) [reply src/components/common/CitiesList/CitiesListSkeleton.jsx:19]: Sure? the component does not accept props and is always rendered the same

Comment (adadychanka) [reply src/components/common/CitiesList/CitiesListSkeleton.jsx]: yes, the result will be the same, but even components without props rerender if a parent has a state changes 

to avoid useless rerenders it might be wrapped into memo

File: src/components/common/CitiesList/CitiesList.jsx [hunk 13]
```diff
@@ -1,21 +1,38 @@
...
+const CitiesList = ({ limit, increaseLimit, showLoadMoreButton }) => {
+  const { data: cities = [], error, isLoading } = usePopularCities();
 
-const CitiesList = ({ cities, limit, increaseLimit, showLoadMoreButton }) => {
   const defaultLimit = 10;
   const citiesLimit = limit || defaultLimit;
-  const displayedCities = cities.slice(0, citiesLimit);
-  const allCitiesDisplayed = citiesLimit >= cities.length;
+  const displayedCities = useMemo(
+    () => cities?.slice(0, citiesLimit) || [],
```
Comment (adadychanka) [inline src/components/common/CitiesList/CitiesList.jsx:14]: you don't need optional chaining or fallback value here due to having a default value while destructing result of `usePopularCities` hook

**Comment by adadychanka** [comment]
Few more UI bugs

image
image
